### ISSUE TITLE
v0.18.0 doesn't find some of my async fixtures

### ISSUE THREAD

**Issue by PerchunPak** [pytest-dev/pytest-asyncio#286]
[Logs from Github Actions.](https://github.com/PerchunPak/PingerBot/runs/5105657604?check_suite_focus=true#step:7:83)
[Code.](https://github.com/PerchunPak/PingerBot/blob/ec09ca75b3a812de007a97e43628a9fb78fc9cfc/tests/test_other_cmd.py#L9)
Keep in mind, that v0.17.x worked normally.

**Comment by seifertm** [comment]
Thanks for the logs and the code example!
Your code is using `@pytest.fixture` to decorate asynchronous fixtures. Starting with v0.17, asynchronous fixtures should be decorated with `@pytest_asyncio.fixture`. (see https://github.com/pytest-dev/pytest-asyncio#async-fixtures)

Could you try changing the decorators on the fixtures in your test and report back?

**Comment by PerchunPak** [comment]
I am using `@pytest.fixture` in other files, and it works fine with `asyncio_mode = auto`.
Logs when I changed all `@pytest.fixture` to `@pytest_asyncio.fixture` in `test_other_cmd`.
```
================================================= test session starts =================================================
platform win32 -- Python 3.10.2, pytest-7.0.0, pluggy-1.0.0
rootdir: C:\Users\perch\Desktop\programming\PingerBot, configfile: pytest.ini, testpaths: tests
plugins: asyncio-0.18.0, cov-3.0.0, forked-1.3.0, rerunfailures-10.2, timeout-2.0.1, xdist-2.4.0
asyncio: mode=auto
collected 56 items

tests\test_cmd_alias.py ......                                                                                   [ 10%]
tests\test_cmd_motd.py .s....                                                                                    [ 21%]
tests\test_cmd_ping.py ......s.                                                                                  [ 35%]
tests\test_cmd_stat.py ........s....                                                                             [ 58%]
tests\test_database.py ...............                                                                           [ 85%]
tests\test_other_cmd.py FFFFFFFF                                                                                 [100%]

====================================================== FAILURES =======================================================
________________________________________ TestOtherCommands.test_help_commands _________________________________________
tests\test_other_cmd.py:34: in test_help_commands
    for field in help_cmd["fields"]:
E   TypeError: 'coroutine' object is not subscriptable
______________________________________ TestOtherCommands.test_about_description _______________________________________
tests\test_other_cmd.py:40: in test_about_description
    assert bot.app_info.description in about.description
E   AttributeError: 'coroutine' object has no attribute 'description'
___________________________________________ TestOtherCommands.test_about_id ___________________________________________
tests\test_other_cmd.py:44: in test_about_id
    assert str(bot.app_info.id) in about.description
E   AttributeError: 'coroutine' object has no attribute 'description'
_________________________________________ TestOtherCommands.test_about_owner __________________________________________
tests\test_other_cmd.py:48: in test_about_owner
    assert str(bot.app_info.owner) == about.fields[0].value
E   AttributeError: 'coroutine' object has no attribute 'fields'
___________________________________ TestOtherCommands.test_about_count_servers_int ____________________________________
tests\test_other_cmd.py:56: in test_about_count_servers_int
    assert isinstance(int(about.fields[1].value), int)
E   AttributeError: 'coroutine' object has no attribute 'fields'
____________________________________ TestOtherCommands.test_about_count_users_int _____________________________________
tests\test_other_cmd.py:64: in test_about_count_users_int
    assert isinstance(int(about.fields[2].value), int)
E   AttributeError: 'coroutine' object has no attribute 'fields'
_____________________________________ TestOtherCommands.test_about_python_version _____________________________________
tests\test_other_cmd.py:68: in test_about_python_version
    assert ".".join(map(str, version_info[:3])) in about.fields[3].value
E   AttributeError: 'coroutine' object has no attribute 'fields'
_________________________________________ TestOtherCommands.test_invite_link __________________________________________
tests\test_other_cmd.py:72: in test_invite_link
    assert f"https://discordapp.com/oauth2/authorize?client_id={bot.app_info.id}&scope=bot&permissions=8" in invite.content
E   AttributeError: 'coroutine' object has no attribute 'content'
================================================== warnings summary ===================================================
tests/test_cmd_alias.py::TestAlias::test_is_added_to_db
  C:\Users\perch\Desktop\programming\PingerBot\tests\conftest.py:35: DeprecationWarning: There is no current event loop
    loop = get_event_loop()

tests/test_other_cmd.py::TestOtherCommands::test_invite_link
  C:\Users\perch\AppData\Local\Programs\Python\Python310\lib\site-packages\_pytest\fixtures.py:1037: RuntimeWarning: coroutine 'TestOtherCommands.help_cmd' was never awaited
    self.cached_result = None

-- Docs: https://docs.pytest.org/en/stable/how-to/capture-warnings.html

---------- coverage: platform win32, python 3.10.2-final-0 -----------
Name                Stmts   Miss Branch BrPart  Cover   Missing
---------------------------------------------------------------
commands.py           216     47     38      0    79%   181-182, 192-193, 207-249, 304-317, 322-340, 345
config.py               4      0      0      0   100%
database.py            48      0      2      0   100%
error_handlers.py      26     26     14      0     0%   1-60
main.py                83     83     12      0     0%   1-130
---------------------------------------------------------------
TOTAL                 377    156     66      0    58%
Coverage HTML written to dir htmlcov

=============================================== short test summary info ===============================================
SKIPPED [1] tests\test_cmd_motd.py:85: фича еще не добавлена
SKIPPED [1] tests\test_cmd_ping.py:108: фича еще не добавлена
SKIPPED [1] tests\test_cmd_stat.py:166: фича еще не добавлена
================================ 8 failed, 45 passed, 3 skipped, 2 warnings in 17.73s =================================
sys:1: RuntimeWarning: coroutine 'TestOtherCommands.about' was never awaited
RuntimeWarning: Enable tracemalloc to get the object allocation traceback
```

**Comment by cnaaq** [comment]
I have the same issue. I figured out something else which is worth mentioning (please see if you can reproduce it as well).
Consider the two fixtures + two test cases:

```py
@pytest.fixture
async def always_false():  # an async fixture 1
    return False

def test_a(always_false):  # a sync test function using the async fixture 1
    assert always_false is False


@pytest.fixture
async def always_true():  # an async fixture 2
    return True

def test_b(always_true):  # a sync test function using the async fixture 2
    assert always_true is True
```

So far nothing new and no apparent difference between the two examples. Both of them fails because of the reported bug.
What I noticed is that fixing one of them, causes both of them to pass! Here:

```py
@pytest.fixture
async def always_false():  # same async fixture 1
    return False

@pytest.mark.asyncio
async def test_a(always_false):  # Fixed!
    assert always_false is False


@pytest.fixture
async def always_true():  # same async fixture 2
    return True

def test_b(always_true):  # same test using fixture 2
    assert always_true is True
```

In this case both test cases pass. Even running the second test case individually pass!

I'm using:
- Python 3.9
- pytest 6.2.5
- pytest-asyncio 0.18

**Comment by PerchunPak** [comment]
@cnaaq, can't reproduce it on

- Windows 11
- Python 3.10.2
- pytest 8.0.0
- pytest-asyncio 0.18.0

Try update your `pytest` or `python`. Also try create virtual environment.

**Comment by seifertm** [comment]
@cnaaq Thanks for the additional information. 

The v0.18.1 release should fix the issue.

@cnaaq @PerchunPak Can you upgrade to v0.18.1 and check if this fixes your problems?

**Comment by cnaaq** [comment]
Yes, thank you very much for the quick fix :)

**Comment by Sebadst** [comment]
> Starting with v0.17, asynchronous fixtures should be decorated with @pytest_asyncio.fixture

@seifertm isn't this a breaking change? Why is it not marked as such? 😕 
I don't even see references about it in the [changelog](https://github.com/pytest-dev/pytest-asyncio/blob/master/CHANGELOG.rst#0170-22-01-13)

**Comment by seifertm** [comment]
@Sebadst pytest-asyncio v0.17 introduced three modes (auto, strict, and legacy) which provide different ways fixtures and tests are discovered by pytest-async. The default mode for v0.17 is _legacy_ which is backwards compatible. Legacy mode raised a warning to notify users about the behaviour change and encouraged them to migrate to either auto or strict mode.

pytest-asyncio v0.19 was released 6 months after 0.17 and changed the behaviour to use strict mode by default. This was marked as a breaking change.

Legacy mode was only removed in v0.20, roughly 10 months after the deprecation. Another breaking change and also announced in the changelog.

I'm sorry that this issue bit you, but claiming that this wasn't marked as a breaking change is simply wrong. I think we went above and beyond to allow for a smooth transition.

I do encourage you to provide feedback on the deprecation process, though. Let us know if there's something we can do to improve the process in the future and we'll definitely consider it.
